### PR TITLE
Autotiles

### PR DESCRIPTION
**Pull request by Semphriss**
Autotiles. Tiles that automatically adapt to their environment :)

Showcase : 
https://www.youtube.com/watch?v=Yb9ZmPOO4fM

### CODE REVIEW

File: src/editor/overlay_widget.cpp [hunk 1]
```diff
@@ -184,11 +232,26 @@ EditorOverlayWidget::draw_rectangle()
...
+        input_autotile( Vector(static_cast<float>(x), static_cast<float>(y)), m_editor.get_tiles()->pos(x_, y_) );
+      } else {
+        input_tile( Vector(static_cast<float>(x), static_cast<float>(y)), m_editor.get_tiles()->pos(x_, y_) );
+      }
     }
   }
 }
 
+bool
+EditorOverlayWidget::check_tiles_for_fill(uint32_t replace_tile, uint32_t target_tile, uint32_t third_tile)
```
Comment (tobbi) [inline src/editor/overlay_widget.cpp:245]: Mark this method `const`

File: src/editor/overlay_widget.cpp [hunk 2]
```diff
@@ -184,11 +232,26 @@ EditorOverlayWidget::draw_rectangle()
...
+{
+  if (autotile_mode) {
+    return AutotileSet::get_tileset_from_tile(replace_tile) == AutotileSet::get_tileset_from_tile(target_tile)
```
Comment (tobbi) [inline src/editor/overlay_widget.cpp:248]: the line width hurts me a bit. Not sure if there's anything that can be done about that.

File: src/supertux/autotile.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+AutotileMask::AutotileMask(uint8_t mask, bool center) :
+  m_mask(mask),
+  m_center(center)
+{
+}
+
+bool
+AutotileMask::matches(uint8_t mask, bool center)
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:31]: Mark this method `const`

File: src/supertux/autotile.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+Autotile::Autotile(uint32_t tile_id, std::vector<AutotileMask*> masks, bool solid) :
+  m_tile_id(tile_id),
+  m_masks(masks),
+  m_solid(solid)
+{
+}
+
+bool
+Autotile::matches(uint8_t num_mask, bool center)
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:46]: Mark this method `const` as well.

File: src/supertux/autotile.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+uint32_t
+Autotile::get_tile_id()
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:59]: Same for this.

File: src/supertux/autotile.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  return m_tile_id;
+}
+
+bool
+Autotile::is_solid()
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:65]: As well as this.

File: src/supertux/autotile.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+AutotileSet::AutotileSet(std::vector<Autotile*> tiles, uint32_t default_tile) :
+  m_autotiles(tiles),
+  m_default(default_tile)
+{
+}
+
+AutotileSet*
+AutotileSet::get_tileset_from_tile(uint32_t tile_id)
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:82]: I'm not sure if this method can be marked `const` as well.

File: src/supertux/autotile.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  return nullptr;
+}
+
+uint32_t
+AutotileSet::get_autotile(uint32_t tile_id,
+    bool top_left, bool top, bool top_right,
+    bool left, bool center, bool right,
+    bool bottom_left, bool bottom, bool bottom_right
+  )
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:111]: But this can be marked `const`

File: src/supertux/autotile.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  }
+
+  return center ? get_default_tile() : 0;
+}
+
+
+
+uint32_t
+AutotileSet::get_default_tile()
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:139]: `const`

File: src/supertux/autotile.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  return m_default;
+}
+
+bool
+AutotileSet::is_member(uint32_t tile_id)
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:145]: `const`

File: src/supertux/autotile.cpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool
+AutotileSet::is_solid(uint32_t tile_id)
+{
```
Comment (tobbi) [inline src/supertux/autotile.cpp:158]: `const`

File: src/supertux/autotile.hpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+  AutotileMask(const AutotileMask&) = delete;
+  AutotileMask& operator=(const AutotileMask&) = delete;
+};
+
+class Autotile final
+{
+public:
+  Autotile(uint32_t tile_id, std::vector<AutotileMask*> masks, bool solid);
+
+  bool matches(uint8_t mask, bool center);
```
Comment (tobbi) [inline src/supertux/autotile.hpp:58]: Mark the corresponding methods `const` here as well.

**Review by tobbi — CHANGES_REQUESTED** [review]
Please fix issues, kthxbye.

File: src/supertux/autotile.cpp (hunk 7 above)
Comment (Semphriss) [reply src/supertux/autotile.cpp:82]: That one isn't only a read; it's also the one that populates m_autotilesets if it is empty.

**Comment by papb** [comment]
Wow this is fantastic!!

**Pull request merged**